src/metabase/models/query/permissions.clj

```diff
@@ -103,6 +104,30 @@
   (binding [api/*current-user-id* nil]
     ((requiring-resolve 'metabase.query-processor.preprocess/preprocess) query)))
 
+(defn- referenced-card-ids
+  "Return the union of all the `::referenced-card-ids` sets anywhere in the query."
+  [query]
+  (let [all-ids (atom #{})]
+    (walk/postwalk
+     (fn [form]
+       (when (map? form)
+         (when-let [ids (not-empty (::referenced-card-ids form))]
+           (swap! all-ids set/union ids)))
+       form)
+     query)
+    (not-empty @all-ids)))
+
+(defn- native-query-perms
+  [query]
+  (merge
+   {:perms/create-queries :query-builder-and-native
```

Comment:
This seems to be "added" up to two times now: once here and once in `legacy-mbql-required-perms` if `native?` is truthy.

Reply:
yeah, because it's in a big `merge` expression that also adds a different value here under different conditions it didn't seem like it was hurting anything. I could have removed the `when native?` part below but it seemed safer to leave it in to be absolutely sure we're using `:query-builder-and-native` if it's native